### ISSUE TITLE
Add javers option

### ISSUE THREAD

**Issue by atomfrede** [deepu105/generator-jhipster-entity-audit#4]
Javers can be an option for auditing and diff of objects.

**Comment by deepu105** [comment]
@atomfrede  Please see how much of the UI can be reused

I originally did filters in UI to search audit by user and from/to date combinations, but I used queryDSL for that and didnt want to introduce that dependency so removed that option. If you have a way of doing that with Spring data JPA plz do so as well :)

**Comment by atomfrede** [comment]
@deepu105 Is it correct, that your module only works with sql? Should we enforce it when using the module (or did I miss something)?

**Comment by deepu105** [comment]
Yes it works only with hibernate and hence sql, i have done that check in
the main generator but might have missed it in the hook sub gen ill check

**Comment by deepu105** [comment]
@atomfrede I have enabled th check in sub generator as well and have released a patch

**Comment by atomfrede** [comment]
I've seen javers has a [spring-boot starter](http://javers.org/documentation/spring-boot-integration/) (incubating at the moment), but it looks quite easy to use. All examples are with mongo, but I think the same auto config mechanism can be used for sql as well. 

So I'll try it out. @bartoszwalacik anything I need to keep in mind when using your starters at the moment?

**Comment by deepu105** [comment]
Ya doing it for spring is easy, need to add some config in config class and
annotate repositories that's all, let me know if you need any code sample,
as i have tried it already once

**Comment by bartoszwalacik** [comment]
@atomfrede javers-spring-boot-starter-mongo is almost ready, give me a while (1-2 days)
and it will be released with JaVers 1.4.1
javers-spring-boot-starter-sql requires bit more work, 1-2 weeks

What is important, we created these starters to be compatible with spring-data starters, all db configuration, transactions is shared, so it's very easy to add JaVers to project which uses spring-data starters

**Comment by bartoszwalacik** [comment]
@atomfrede JaVers Spring Boot starter for MongoDB is released, see
http://javers.org/documentation/spring-boot-integration/
I am waiting for your feedback.

**Comment by deepu105** [comment]
Good news

**Comment by atomfrede** [comment]
:+1: Wil try it today.

**Comment by bartoszwalacik** [comment]
Well, mongo 2.x is quite old, mongo client 3.0 is backward compatible wit mongo 2.x servers. Is upgrade to 3.0 an option for you?

**Comment by atomfrede** [comment]
Yes I'll try it. Mongeez has a new version supporting mongo-client 3.x. Will post updates here.

**Comment by atomfrede** [comment]
I got it working. Need to do a small PR on the main generator updating mongeez, mongo javadriver and the database configuration.

**Comment by deepu105** [comment]
Nice man

**Comment by bartoszwalacik** [comment]
@atomfrede let me know when your module will be available for testing

**Comment by atomfrede** [comment]
I can setup a sample repository (to get some feedback) while implementing it in the generator module.

**Comment by atomfrede** [comment]
@deepu105 In the entity subgenerator I need the choosen auditing framework, whats the best place to store it? Or is it already stored somewhere and I didn't find it?

**Comment by deepu105** [comment]
@atomfrede currently im storing it as `enableEntityAudit` a boolean in the respective entity.json file, you can change it to store a string as required.

About PR that is up to you, plz do as you feel comfortable, if you want to PR and add on to it progressively thats fine as well

**Comment by atomfrede** [comment]
A little status update:

* executing the main audit generator works fine (just a little fine tuning needed for annotating repositories)
* audit resource not yet able to provide javers audits (should be easy to implement it such that the ui can handle javers changelogs)
* main problem: When generating a new entity, I must know the selected audit framework in the callback, need to find a place where to store it

**Comment by deepu105** [comment]
Fred, I would advice storing in in the entity json or as a yo-rc param from
the module itself so that it can be managed by the module

**Comment by atomfrede** [comment]
@bartoszwalacik In order to reuse @deepu105 ui for viewing changes I need to find 
* all entities that are audited (this is a minor point, I can handle this via a configuration or so)
* need to find the snaphot (for a specific entity) and a specific commit id (e.g. get the state of the entity before or at commit x)

Is this somehow possible? I didn't find something in the documentation.

**Comment by bartoszwalacik** [comment]
There are no such queries yet.
Query for all audited entities would require some API changes. Query with CommitId better fits to existing API.

new queries could  look like this:

```
Snapshot findSnapshot(Class entity, Object id, CommitId commitId);
List<Snapshot> findSnapshots(Class entity, CommitId commitId);
List<Snapshot> findSnapshots(CommitId commitId);
``` 
Which one do you need?

**Comment by atomfrede** [comment]
As I said query for all audited entites would be a nice to have, I can do that differently.

All three look quite helpful, for for the concrete usecase I think the first one would be great.

**Comment by deepu105** [comment]
All audited entities are registered in a javers table, i created entity for
that table and did get all to get all audited entity names, when i was
trying out javers earlier

**Comment by bartoszwalacik** [comment]
CommitId filter is released in JaVers v 1.4.5
see http://javers.org/documentation/jql-examples/#commit-id-filter

**Comment by atomfrede** [comment]
Thanks @bartoszwalacik! You and your team are awesome!

**Comment by bartoszwalacik** [comment]
@atomfrede nice to hear that! Let us know, if we can help with testing or improving new jhipster-javers features. We'd like to see how it all works together.

**Comment by atomfrede** [comment]
@bartoszwalacik In deed, maybe I have another thing ;) Is it possible to filter by commit version too?

**Comment by bartoszwalacik** [comment]
you mean snapshot.version? Not yet, snapshot versions are quite new feature ...

**Comment by atomfrede** [comment]
Yes. I've seen it to late, but version is exactly what I need, as it is not global. So when I have version x the previous snapshot is x - 1, which is not the case for the commit id.

**Comment by bartoszwalacik** [comment]
looks doable, I've created the issue for that https://github.com/javers/javers/issues/322

**Comment by bartoszwalacik** [comment]
hi, https://github.com/javers/javers/issues/322 is done, check JaVers v. 1.4.7

**Comment by deepu105** [comment]
That was fast man :)

Thanks & Regards,
Deepu

**Comment by atomfrede** [comment]
:+1:

**Comment by deepu105** [comment]
*@atomfrede let me know when its ready to merge*

Thanks & Regards,
Deepu

**Comment by atomfrede** [comment]
I'll try to update my pr today. So I guess you can merge it today and test it.

**Comment by atomfrede** [comment]
@deepu105 I think you can test it now.

**Comment by deepu105** [comment]
closed via #9

**Comment by raymondttitusgt** [comment]
I have issue making javers entity audit to work in jhipster
